File: leetcode/dp/2378_Choose_Edges_to_Maximize_Score_in_a_Tree/codes/dp.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    long long maxScore(vector<vector<int>>& edges) {
        int n = edges.size();
        vector<vector<pair<int,int>>> g(n);
        for (int i = 1; i < n; i++) {
            int p = edges[i][0], w = edges[i][1];
            g[p].push_back({i, w});
        }
        vector<array<long long,2>> dp(n, {LLONG_MIN, LLONG_MIN});

        function<long long(int,int)> dp = [&](int u, int used) -> long long {
            long long &res = dp[u][used];
            if (res != LLONG_MIN) return res;
            if (g[u].empty()) return res = 0;
            long long ans = 0;
            for (auto &e : g[u]) ans += dp(e.first, 0);

            if (!used) {
                for (auto &e : g[u]) {
                    int v = e.first, w = e.second;
                    if (w <= 0) continue;
                    long long tmp = 0;
                    for (auto &ee : g[u]) {
                        if (ee.first == v) tmp += w + dp(ee.first, 1);
                        else tmp += dp(ee.first, 0);
                    }
                    ans = max(ans, tmp);
                }
            }
            return res = ans;
        };

        return dp(0, 0);
    }
};

File: leetcode/dp/2378_Choose_Edges_to_Maximize_Score_in_a_Tree/codes/dp.py
from collections import defaultdict
from functools import lru_cache

class Solution:
    def maxScore(self, edges: List[List[int]]) -> int:
        n = len(edges)
        g = defaultdict(list)
        for i, (p, w) in enumerate(edges[1:], 1):
            g[p].append((i, w))

        @lru_cache(None)
        def dp(u, used):
            res = sum(dp(v, False) for v, _ in g[u])
            if not used:
                for v, w in g[u]:
                    if w <= 0: 
                        continue
                    cur = w + dp(v, True) + sum(dp(x, False) for x, _ in g[u] if x != v)
                    res = max(res, cur)
            return res
        
        return dp(0, False)
